### PR TITLE
MGMT-11265: add platform integration enhancement

### PR DESCRIPTION
**Pull request by vrutkovs**
This enhancement describes existing "Integrate with vSphere/BM" feature and describes an updated proposal to extend the API to support more providers

## List all the issues related to this PR

- [ ] New Feature 
- [x] Enhancement 
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [x] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- Are the title and description (in both PR and commit) meaningful and clear?
- Is there a bug required (and linked) for this change?
- Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: docs/enhancements/platform-integration.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+- Define requirements for other components - most importantly, openshift-installer - required to be fulfilled to allow assisted-installer team add more integrations in the future.
+
+### Non-Goals
+
+- Integrating any relevant [Openshift installer types](https://github.com/openshift/installer/tree/master/pkg/types) automatically - this would require vendoring a particular openshift-installer version, while assisted-installer aims to support multiple versions
+
+## Proposal
+
+- Extend Provider interface to list features available in particular provider
+- Provider features should be version-dependent (i.e. Nutanix integration available since Openshift 4.11+)
```
Comment (gamli75) [inline docs/enhancements/platform-integration.md:33]: each provider might have different limitations such as working/ doesn't work on SNO, supporting UMN/ CMN, static IP

File: docs/enhancements/platform-integration.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+- Document requirements for openshift-installer
+
```
Comment (gamli75) [inline docs/enhancements/platform-integration.md:35]: each provider might have different validations

File: docs/enhancements/platform-integration.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+authors:
+  - "@vrutkovs"
+creation-date: 2022-07-26
+last-updated: 2022-07-26
+---
+
+# Platform integration
+
+## Summary
+Assisted installed has historically been used to configure baremetal clusters. Eventually, users have grown accustomed to use it for virtual machines in the cloud, as it helped them to validate networking settings. The installed cluster however could not make use of cloud features (like scaling, volumes or loadbalancers) as it was installed with platform: none. Later on basic platform integration was added, described in [Define Providers Interface](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/add-external-providers.md).
```
Comment (omertuc) [inline docs/enhancements/platform-integration.md:12]:
```suggestion
Assisted installer has historically been used to configure baremetal clusters. Eventually, users have grown accustomed to use it for virtual machines in the cloud, as it helped them to validate networking settings. The installed cluster however could not make use of cloud features (like scaling, volumes or loadbalancers) as it was installed with platform: none. Later on basic platform integration was added, described in [Define Providers Interface](https://github.com/openshift/assisted-service/blob/master/docs/enhancements/add-external-providers.md).
```

File: docs/enhancements/platform-integration.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+## Motivation
+
+Adding other providers for integration with assisted installer should be easy and customizable, including various UX/UI changes for flexibility.
+
+### Goals
+
+- Define API changes necessary to support various providers
+- Define API for UI to allow being flexible about provider features
+- Define requirements for other components - most importantly, openshift-installer - required to be fulfilled to allow assisted-installer team add more integrations in the future.
+
```
Comment (omertuc) [inline docs/enhancements/platform-integration.md:25]: Is migrating existing platform integrations (metal, vsphere, nutanix) to use the new API a goal, or a non-goal?

Comment (vrutkovs) [reply docs/enhancements/platform-integration.md:26]: Good point, added as explicit goal

**Comment by oourfali** [comment]
@gamli75 @eliorerz can you please review as well?

**Comment by oourfali** [comment]
CC @gamli75

File: docs/enhancements/platform-integration.md [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+Implementing provider-specific flows would extend our testing matrix and may confuse users. The mitigation may be additional CI tests and feature-gating UI changes.
+
+## Design Details
+
+### Closed Questions
+
+### UI Impact
+
+UI needs to use `usages` API to selectively enable or disable UI features.
```
Comment (gamli75) [inline docs/enhancements/platform-integration.md:123]: 'usages'? I think it should be 'feature supported API

**Review by gamli75 — COMMENTED** [review]
I think we should add something about 
1. day 2 operations - setting credentials, scale-out clusters 
2. limitations, such as UMN, installing OLM operators

File: docs/enhancements/platform-integration.md (hunk 5 above)
Comment (vrutkovs) [reply docs/enhancements/platform-integration.md:126]: Good catch, updated

**Comment by vrutkovs** [comment]
>I think we should add something about

>  day 2 operations - setting credentials, scale-out clusters

Added a note about post-installation procedure. At minimum assisted-service should have an API for UI to get KB link. Let me know if we want to expand that integration

>   limitations, such as UMN, installing OLM operators

Its mentioned in suggested features:
```
              enum:
              - `ENABLE_CNV`
              - `ENABLE_ODF`
```

**Review by gamli75 — APPROVED** [review]

**Pull request merged**